base/MAPL_Resource.F90

```diff
@@ -0,0 +1,411 @@
+         if (default_is_present .and. .not. label_is_present) then
+            select type(default)
+            type is(integer(real32))
+               vals = default
+            end select
+         else
+            call ESMF_ConfigGetAttribute(config, valuelist = vals, count = count, label = label_to_use, _RC)
+         end if
+      type is (real(real64))
+         if (default_is_present .and. .not. label_is_present) then
+            select type(default)
+            type is(integer(real64))
+               vals = default
+            end select
+         else
+            call ESMF_ConfigGetAttribute(config, valuelist = vals, count = count, label = label_to_use, _RC)
+         end if
+      type is(character(len=*))
+         if (default_is_present .and. .not. label_is_present) then
+            select type(default)
+            type is(character(*))
+               vals = default
+            end select
+         else
+            call ESMF_ConfigGetAttribute(config, valuelist = vals, count = count, label = label_to_use, _RC)
+         end if
+      type is(logical)
+         if (default_is_present .and. .not. label_is_present) then
+            select type(default)
+            type is(logical)
+               vals = default
+            end select
+         else
+            call ESMF_ConfigGetAttribute(config, valuelist = vals, count = count, label = label_to_use, _RC)
+         end if
+         class default
+         _FAIL( "Unsupported type")
+      end select
+
+      _RETURN(ESMF_SUCCESS)
+
+   end subroutine MAPL_GetResource_config_array
+
+   subroutine print_resource(printrc, label, val, default, rc)
+      integer, intent(in) :: printrc
+      character(len=*), intent(in) :: label
+      class(*), intent(in) :: val
+      class(*), optional, intent(in) :: default
+      integer, optional, intent(out) :: rc
+
+      character(len=:), allocatable :: val_str, default_str, output_format, type_str, type_format
+      type(StringVector), pointer, save :: already_printed_labels => null()
+
+      if (.not. associated(already_printed_labels)) then
+         allocate(already_printed_labels)
+      end if
+
+      ! Do not print label more than once
+      if (.not. vector_contains_str(already_printed_labels, trim(label))) then
+         call already_printed_labels%push_back(trim(label))
+      else
+         return
+      end if
+
+      select type(val)
+      type is(integer(int32))
+         type_str = "'Integer*4 '"
+         type_format = '(i0.1)'
+         val_str = intrinsic_to_string(val, type_format)
+         if (present(default)) then
+            default_str = intrinsic_to_string(default, type_format)
+         end if
+      type is(integer(int64))
+         type_str = "'Integer*8 '"
+         type_format = '(i0.1)'
+         val_str = intrinsic_to_string(val, type_format)
+         if (present(default)) then
+            default_str = intrinsic_to_string(default, type_format)
+         end if
```

Comment:
The FPP macro here will probably need 3 arguments:  `type, type_str, and type_format`.

Reply:
Done.